### DISCUSSION TITLE
[ios] [help] Any idea what this rendering artifact is?

### DISCUSSION THREAD

**Discussion by cpheinrich** [google/filament#6353] [Q&A] [answered]
Hello, I am having trouble fixing a rendering artifact that we are seeing on our iOS Filament viewer (but not seeing it on web or android). I have also loaded the glb into Filament's sample gltf viewer and didn't see the issue there either. I've tried configuring various anti aliasing and render quality settings to no avail. I don't expect a bug fix, just want someone to have a look and see if you can tell it is something obvious we are missing  (I compared to filament's sample gltf viewer but couldn't find anything)

![pixelation](https://user-images.githubusercontent.com/12475454/205792801-cc0130bf-9391-4ba6-bf13-de2afc4158d2.jpeg)

On our other clients it looks like this. Notice the lack of pixelation on the text and the face of the teddy bear. 

![Screen Shot 2022-11-21 at 12 26 09 PM](https://user-images.githubusercontent.com/12475454/205792868-1e3cdc9e-1865-4b14-a422-870312bcab9e.png)

**Comment by romainguy** [comment, marked as answer]
It looks like an issue with the filtering mode of the sampler. Could be a bug on our end but it also looks suspiciously like issue #6344. @bejado, maybe there's a bug with spirv-opt and MSL?

**Reply by cpheinrich** [reply]
Ok, thanks, we will follow up on these leads and see if we can make progress and will update this thread.

**Reply by romainguy** [reply]
We'll be able to take a closer look at this when @bejado is back.

**Reply by cpheinrich** [reply]
Ok, also one other quick question -- when the materials are compiled, is there any difference between how they are compiled for iOS vs Mac (I would assume not since they both use Metal backend), since it's worth pointing out that we also have a Mac viewer and this issue is not present there, so it really appears to only be present on iOS from what I can tell. The issue does look similarish to #6344 though since the artifact essentially looks like aliasing that is only present in certain parts of the scene

**Reply by romainguy** [reply]
It's not aliasing, it's absence of texture filtering. Filament uses slightly different shaders on mobile vs desktop, mostly for performance reasons but they are otherwise close. The problem is also that the drivers and GPUs are different, so we might be hitting a combo of optimization + driver causing the problem.

**Reply by cpheinrich** [reply]
Ok I can confirm that disabling material optimization with `set(MATC_OPT_FLAGS -gd) ` even in Release builds fixes the issue. We are going to just always set this flag when building for iOS in our fork

**Reply by romainguy** [reply]
Glad to hear you have a workaround. This indicates a bug in the [spirv-opt](https://github.com/KhronosGroup/SPIRV-Tools) library most likely. We'll try to upgrade to a newer version and hope it fixes the problem.

**Reply by jeanlemotan** [reply]
I've seen something similar when using fp16 for UVs and rendering geometry with large texcoords (>16 or so). @cpheinrich could you check the UVs to see if they are large?

**Reply by bejado** [reply]
I'm able to reproduce this. I can workaround it by commenting-out [these lines](https://github.com/google/filament/blob/main/libs/filamat/src/GLSLPostProcessor.cpp#L543-L547), which turn relaxed precision operations into fp16. It's unclear if this is a bug in spirv-opt, or a precision issue.

**Comment by cpheinrich** [comment]
See thread for full answer to the issue.

**Comment by bejado** [comment]
After some more investigation, I'm almost positive this is due to a SPIRV-opt/cross bug affecting texture coordinate precision and nothing to do with texture sampling. Thanks to @jeanlemotan for the lead.

Some background info: MSL (Metal Shading Language) has no concept of `highp` or `mediump`. Instead, it only deals with concrete precision types, like `float` and `half`. When material optimizations are enabled, GLSL types marked `lowp` or `mediump` are translated to MSL's `half` type.

The quantification you're seeing is caused by the use of half precision UV coordinates. However, if you look at our gltfio materials, you'll see we're careful to mark UV coordinates as `highp`:
```
// base.mat.in

highp float2 uvs[2];
uvs[0] = getUV0();
uvs[1] = getUV1();
...
if (materialParams.baseColorIndex > -1) {
    highp float2 uv = uvs[materialParams.baseColorIndex];
    material.baseColor *= texture(materialParams_baseColorMap, uv);
}
```
So what gives?

I've tracked this down to a (possible) bug in the SPIR-V toolchain (either with SPIRV-opt or SPIRV-cross). Here's a simplified GLSL shader:

```
layout(location = 10) in highp vec4 vertex_uv01;
...
fragColor = texture(materialParams_baseColorMap, vertex_uv01.xy);
```

Even though `vertex_uv01` is marked `highp`, this gets translated to the following MSL:
```
out.fragColor = materialParams_baseColorMap.sample(materialParams_baseColorMapSmplr, float2(half4(in.vertex_uv01).xy));
```
Note the unnecessary `half4` conversion. A similar thing happens in our gltfio materials, and probably any material compiled with optimizations.

The best workaround I could find is to mark the sampler in the material as `highp`. This "tricks" the SPIR-V toolchain into keeping the texture coordinates at full precision.

Here's an exaggerated example screenshot using our gltf-viewer app. The glTF model has texture coordinates ranging from 100.0 to 101.0 and a texture wrap mode set to repeat. The left side is the state of things today, and the right side is after changing the precision of the base color sampler to `highp`, like so:

```
diff --git a/libs/gltfio/materials/base.mat.in b/libs/gltfio/materials/base.mat.in
index da62ce9f8..7f9de3d37 100644
--- a/libs/gltfio/materials/base.mat.in
+++ b/libs/gltfio/materials/base.mat.in
@@ -18,7 +18,7 @@ material {
         // Base Color
         { type : int, name : baseColorIndex },
         { type : float4, name : baseColorFactor },
-        { type : sampler2d, name : baseColorMap },
+        { type : sampler2d, name : baseColorMap, precision: high },
         { type : mat3, name : baseColorUvMatrix, precision: high },
```

![uv-precision-comparison](https://user-images.githubusercontent.com/5298046/209001040-8fdaa9a5-f3f0-414e-a726-51066fcd073c.png)

I filed a bug with SPIRV-Tools: https://github.com/KhronosGroup/SPIRV-Tools/issues/5044
